### PR TITLE
Python 2+3 Compatibility

### PR DESCRIPTION
**Pull request by harelba**
Including lots of changes to the test suite, but semantics is almost not changed. The idea was to minimize the semantic changes to the test code, so we'll still be able to trust them.

for py2, all tests are passing, for py3 the following still fails:
* Float precision has changed, so it breaks some tests in py3.
* Support for UTF8 with BOM is still not working
* Universal newline support is not working as well
* `-f` parameter format has been changed for now to `.format()` format instead of `%` format. This is obviously a breaking change, but the idea was to see if everything works well, and then decide if this feature is worth the extra effort of leaving it compatible. In general, it's a "bad feature" in the sense that it exposes the implementation to the interface level, and actually I'm not that sure that it's being used enough. Anyway, if the decision would be that this has to remain compatible, then I would implement something to bridge the gap.

I really appreciate everyone's attempts to contribute to this. However, due to the delicate nature of encoding/decoding changes in py3 and the amount of code related to it in q and in the test suite, I have tried to make a concentrated effort in order to solve this in one coherent pass and not as pure syntactic replacements. 

Here's a summary of what I did:
* In the tests themselves - All string data is now six.b, since this was the way that py2 treated it, and this is what was in mind when writing the tests. This also allows knowing exactly when is written to files/etc., with minimum intervention from python itself. As i mentioned above, making no/easy-to-reason changes in the tests allows to continue to treat them as trustworthy.
* Also for tests - using `open` in some of the cases was not enough, since the test assumed that there's an encoding to the data being written to the file. For these cases, i replaced `file` with `codecs.open(...encoding='utf-8')`. The reason is that when writing the tests, the "strings" (e.g. byte arrays) that i've written to these files were implicitly in mind UTF-8 in terms of encoding. Perhaps this requires setting t
he python source encoding explicitly to utf-8 as well in order to be more robust to running the tests in non utf-8 environments.
* All % formatting was converted to `.format()`, since it's unicode behaviour was more predictable. This includes for now the breaking behaviour of the `-f` parameter.
* `run_command` in the test suite is returning un-encoded bytes. Not sure this is the best behaviour i would expect from Popen in py3, but this actually made it easier to eliminate the hidden layer of stdout/stde
rr encoding, and leave most of the tests with similar semantics (doing six.b comparisons instead of string comparisons, leaving the old intended semantics of the tests identical to what it was).
* I've added a simpler (hackish) way to debug specific tests - When running the test suite with a `<test-class>.<test-name>` or `<test-class>` only the relevant tests are being run. Also, running the tests with a
n additional `-v` parameter spits out the q command that was executed, and the results from run_command, and prevents deleting the temporary files that were created as part of the test. This allows easy multiple
reruns of the actual q command line that was used in the test.
* There are a few places where I've postponed "merging" the code between py2 and py3 in order to expedite the learning curve of the encoding related issues. The idea was to get that out of the way (not too many lines of code), and to address the merge later:
  1. The encoded_csv_reader
  2. The encoding of column names arriving from the DB driver
  3. The encoding of STDOUT

@cclauss and anyone interested - I would appreciate any input you have on this PR, and on the choices I've made.

Btw, added python 3.7 to travis to see if things work there as well.

### CODE REVIEW

**Comment by cclauss** [comment]
travis-ci/travis-ci#9069 Python 3.7 on Travis needs to run on __dist: xenial__ see https://github.com/jupyter/jupyter/pull/385/files

**Comment by harelba** [comment]
just saw this as well 😄 pushing the fix now

File: bin/q [hunk 1]
```diff
@@ -43,13 +43,17 @@ import codecs
...
 import csv
 import hashlib
 import uuid
-import cStringIO
 import math
+import six
+
+if six.PY3:
+    long = int
+    unicode = six.text_type
```
Comment (cclauss) [inline bin/q]: This approach does not follow https://docs.python.org/3/howto/pyporting.html#use-feature-detection-instead-of-version-detection  Also __unicode(x, encoding)__ will fail in Python 3 and you have an instance of that in this file.

File: bin/q [hunk 2]
```diff
@@ -418,12 +424,12 @@ class Sql(object):
             qtable_name] = effective_table_name
 
     def get_effective_sql(self,original_names=False):
-        if len(filter(lambda x: x is None, self.qtable_name_effective_table_names)) != 0:
+        if len(list(filter(lambda x: x is None, self.qtable_name_effective_table_names))) != 0:
```
Comment (cclauss) [inline bin/q]: list comprehension is easier to understand.

File: bin/q [hunk 3]
```diff
@@ -641,7 +647,7 @@ class TableColumnInferer(object):
         counts = {}
         for column_count in column_count_list:
             counts[column_count] = counts.get(column_count, 0) + 1
-        return ", ".join(["%s rows with %s columns" % (v, k) for k, v in counts.iteritems()])
+        return six.u(", ").join([six.u("{} rows with {} columns".format(v, k)) for k, v in six.iteritems(counts)])
```
Comment (cclauss) [inline bin/q]: six.u() and six.b() and not widely used these daze with u"string" and b"string" being supported on all supported versions of Python.

File: bin/q [hunk 4]
```diff
@@ -611,7 +617,7 @@ class TableColumnInferer(object):
                     # in relaxed mode, add columns to fill the missing ones
                     self.header_row = self.header_row + \
                         ['c%s' % (x + len(self.header_row) + 1)
-                         for x in xrange(self.column_count - len(self.header_row))]
+                         for x in range(self.column_count - len(self.header_row))]
```
Comment (cclauss) [inline bin/q]: If __x__ is really large __range(x)__ on Python 2 has memory overhead but if x is small who cares.  In general I use __from six import xrange__ instead.

**Comment by harelba** [comment]
@cclauss do you happen to know any best-practices regarding the float/double precision changes between py2 and py3?

File: .travis.yml [hunk 5]
```diff
@@ -2,9 +2,11 @@ language: python
 python:
   - "2.7"
   - "3.6"
+  - "3.7"
 matrix:
   allow_failures:
     - python: "3.6"
-install: pip install flake8
+    - python: "3.7"
+install: pip install -r requirements.txt
```
Comment (cclauss) [inline .travis.yml]: You should lint your tests too.

File: bin/q [hunk 6]
```diff
@@ -179,7 +182,10 @@ class Sqlite3DB(object):
             self.cursor.execute(q)
             if self.cursor.description is not None:
                 # we decode the column names, so they can be encoded to any output format later on
-                query_column_names = [c[0].decode('utf-8') for c in self.cursor.description]
+                if six.PY2:
+                    query_column_names = [unicode(c[0],'utf-8') for c in self.cursor.description]
+                else:
+                    query_column_names = [c[0] for c in self.cursor.description]
```
Comment (cclauss) [inline bin/q]: What about __c[0].decode('uft-8')__ in both Py2 and Py3?

File: bin/q [hunk 7]
```diff
@@ -1272,7 +1306,7 @@ class QTextAsData(object):
         self._load_data(filename,input_params,stop_after_analysis=stop_after_analysis)
 
     def load_data_from_string(self,filename,str_data,input_params=QInputParams(),stop_after_analysis=False):
-        sf = cStringIO.StringIO(str_data)
+        sf = six.StringIO(str_data)
```
Comment (cclauss) [inline bin/q]: __six.cStringIO()__ might be faster on Python2 but I doubt it matters here.

**Comment by harelba** [comment]
Thanks a lot @cclauss ! i'll perform the fixes.

Also, forget about the float/double thing - It's actually a bug in q because of __str__() vs __repr__() differences that does not happen in py3... I believe I'll just fix the bug itself.

**Comment by cclauss** [comment]
https://github.com/harelba/q/blob/master/test/test-suite#L1999 __file()__ was removed in Python 3 (there are other issues too.

**Comment by harelba** [comment]
> https://github.com/harelba/q/blob/master/test/test-suite#L1999 **file()** was removed in Python 3 (there are other issues too).

The link you've send is pointing to master - i haven't merged the branch yet.

File: bin/q [hunk 8]
```diff
@@ -778,7 +781,7 @@ class MaterializedFileState(object):
         if self.encoding == 'utf-8-sig' and self.lines_read == 0 and not self.skipped_bom:
             try:
                 BOM = self.f.read(3)
-                if BOM != '\xef\xbb\xbf':
+                if BOM != six.b('\xef\xbb\xbf'):
```
Comment (cclauss) [inline bin/q]: b'\xef\xbb\xbf' works.

Comment (harelba) [reply bin/q]: saw your important note regarding six.b and six.u not being used anymore. Figured that I'll change everything to be consistent, and then do a find-replace sweep and change all of those to the regular b and u.

File: bin/q (hunk 6 above)
Comment (harelba) [reply bin/q]: the cursor returns byte strings for py2, and already-decoded strings for py3.

File: .travis.yml (hunk 5 above)
Comment (harelba) [reply .travis.yml]: 👍

File: bin/q (hunk 4 above)
Comment (harelba) [reply bin/q]: 👍 decided not to do it for now, like you said in most cases this should be a non issue. I'll do some benchmarks later on about the difference between py2 and py3 behavior, and fix this if needed.

File: bin/q (hunk 3 above)
Comment (harelba) [reply bin/q]: Thanks. Will do a complete sweep over all of those after everything works.

File: bin/q (hunk 2 above)
Comment (harelba) [reply bin/q]: true. decided not to change it now, will do later on.

File: bin/q (hunk 1 above)
Comment (harelba) [reply bin/q]: removed the unicode(x, encoding) code, thanks for noticing. As for the long/unicode checking, do you happen to have a suggestion on how to do that using feature-detection?

**Comment by harelba** [comment]
everything now works, with some minor breaking change which i'm not sure is critical - py3's universal newline behavior is such that it automatically succeeds to detect it even if it's not explicitly provided, so the breaking change is that even if someone doesn't provide the `-U` flag, things will go smoothly.

Odd - now i'm seeing that python 3.6 passed for universal newlines, but 3.7 fails on it. checking.

File: bin/q [hunk 9]
```diff
@@ -780,7 +778,11 @@ class MaterializedFileState(object):
         # multiple people.
         if self.encoding == 'utf-8-sig' and self.lines_read == 0 and not self.skipped_bom:
             try:
-                BOM = self.f.read(3)
+                if six.PY2:
+                    BOM = self.f.read(3)
+                else:
+                    BOM = self.f.buffer.read(3)
+
```
Comment (cclauss) [inline bin/q]: You don't like the ternary style?  __BOM = self.f.read(3) if six.PY2 else self.f.buffer.read(3)__

**Comment by cclauss** [comment]
Do you have any performance benchmarks for comparing Py2 and Py3?

**Comment by harelba** [comment]
@cclauss actually i did some. Nothing too scientific yet, but it seems that py3 is significantly better than py2. I've also done similar comparisons with [textql](https://github.com/dinedal/textql), to see the differences between them in terms of performance.

Interesting results, i'm still working on it, to make sure that results are consistent. Will publish when available.

**Comment by harelba** [comment]
Code is now fully compatible with both python versions, with a possible significant performance improvement. The next challenge is now to create proper packaging for it. I'll work on this in a separate PR after merging. Afterwards, I'll do some project structure changes in order to export q as a python module as well.

**Comment by harelba** [comment]
Regarding the performance gains, I'm performing a benchmark and will update once results are final.

**Comment by harelba** [comment]
Merging this. Note that there's one "breaking change" in terms of packaging - One cannot anymore download the q python file straight from the master and just run it. The reason is that now external python dependencies are required. I am deprecating the "one-exectuable-file" option, so I believe that this is a reasonable decision. The plan is in line with exposing q as a python module as well, which i'll address soon.

**Pull request merged**